### PR TITLE
Introduce Adoption APIs

### PR DESCRIPTION
**Pull request by thetechnick**
This PR starts work on supporting gradual handovers of operands between operator versions within a roll-out process.
`Adoption` is the first API, which is just adding labels to operands without any labels defined.
For the Handover process, this API is used to route freshly created objects to an operator.

e.g.:
my-operator v1 is deployed and only listening for `MyObjects` with a `version: v1` label.
Now new objects are not serviced until they get this label added.

By creating an `Adoption` object with static label strategy, new objects are automatically given this label and are serviced by the Operator.

my-operator v2 can later be installed alongside v1 and operand ownership transferred by changing the label value from v1 to v2.

### CODE REVIEW

File: config/crds/coordination.package-operator.run_adoptions.yaml [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+          status:
+            default:
+              phase: Pending
+            description: AdoptionStatus defines the observed state of an Adoption.
+            properties:
+              conditions:
+                description: Conditions is a list of status conditions ths object
+                  is in.
+                items:
+                  description: "Condition contains details for one aspect of the current
```
Comment (eqrx) [inline config/crds/coordination.package-operator.run_adoptions.yaml:123]: yaml multiline string for readability maybe? I mean, in the end nobody reads this in code .... except that I just did :D

File: config/crds/coordination.package-operator.run_clusteradoptions.yaml [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+            default:
+              phase: Pending
+            description: ClusterAdoptionStatus defines the observed state of an ClusterAdoption.
+            properties:
+              conditions:
+                description: Conditions is a list of status conditions ths object
+                  is in.
+                items:
+                  description: "Condition contains details for one aspect of the current
+                    state of this API Resource. --- This struct is intended for direct
```
Comment (eqrx) [inline config/crds/coordination.package-operator.run_clusteradoptions.yaml:124]: See above

File: config/static-deployment/1-coordination.package-operator.run_adoptions.yaml [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+              phase: Pending
+            description: AdoptionStatus defines the observed state of an Adoption.
+            properties:
+              conditions:
+                description: Conditions is a list of status conditions ths object
+                  is in.
+                items:
+                  description: "Condition contains details for one aspect of the current
+                    state of this API Resource. --- This struct is intended for direct
+                    use as an array at the field path .status.conditions.  For example,
```
Comment (eqrx) [inline config/static-deployment/1-coordination.package-operator.run_adoptions.yaml:125]: See above

File: config/static-deployment/1-coordination.package-operator.run_clusteradoptions.yaml [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+            description: ClusterAdoptionStatus defines the observed state of an ClusterAdoption.
+            properties:
+              conditions:
+                description: Conditions is a list of status conditions ths object
+                  is in.
+                items:
+                  description: "Condition contains details for one aspect of the current
+                    state of this API Resource. --- This struct is intended for direct
+                    use as an array at the field path .status.conditions.  For example,
+                    \n type FooStatus struct{ // Represents the observations of a
```
Comment (eqrx) [inline config/static-deployment/1-coordination.package-operator.run_clusteradoptions.yaml:126]: Even more see above

File: internal/controllers/adoption/adoption_controller.go [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+)
+
+type reconciler interface {
+	Reconcile(ctx context.Context, adoption genericAdoption) (ctrl.Result, error)
+}
+
+// Generic reconciler for both Adoption and ClusterAdoption objects.
+type GenericAdoptionController struct {
+	newAdoption genericAdoptionFactory
+
```
Comment (eqrx) [inline internal/controllers/adoption/adoption_controller.go:25]: Empty line supposed to be here?

**Comment by thetechnick** [comment]
Replaced by #223

**Pull request closed without merging**